src/core_modules/capture-core/components/WidgetProfile/WidgetProfile.component.js

```diff
@@ -54,25 +56,46 @@ export const WidgetProfile = ({ teiId, programId }: Props) => {
         return convertClientToView(convertToClientValue, valueType);
     };
 
-    const mergeAttributes = () => {
-        const { programs: { programTrackedEntityAttributes } } = programsData;
-        const { trackedEntityInstances: { attributes } } = trackedEntityInstancesData;
 
-        return programTrackedEntityAttributes
-            .filter(item => item.displayInList)
-            .reduce((acc, curr) => {
-                const foundAttribute = attributes
-                    .find(item => item.attribute === curr.trackedEntityAttribute.id);
+    useEffect(() => {
+        const getListAttributes = async () => {
+            if (programsData && trackedEntityInstancesData) {
+                const { programs: { programTrackedEntityAttributes } } = programsData;
+                const { trackedEntityInstances: { attributes } } = trackedEntityInstancesData;
 
-                acc.push({
-                    reactKey: curr.trackedEntityAttribute.id,
-                    key: curr.trackedEntityAttribute.displayName,
-                    value: formatValue(foundAttribute?.value, foundAttribute?.valueType),
-                });
+                setListAttributes(
+                    await programTrackedEntityAttributes
+                        .filter(item => item.displayInList)
+                        .reduce(async (promisedAcc, { trackedEntityAttribute: { id, displayName, optionSet } }) => {
+                            const foundAttribute = attributes
+                                .find(item => item.attribute === id);
+                            let value;
```

Comment:
Hey, 
I feel that the complexity of this `reduce` is high. Can the whole `if-else` logic where you set the `value` be extracted to another function? 
Thanks!

Reply:
hey, I feel the complexity here is valid (eslint doesn't complain) and moving to other function may make the flow more difficult to follow